### ISSUE TITLE
Error: An I/O error has occurred while writing a response message entity to the container output stream

### ISSUE THREAD

**Issue by ashwinhs** [prestodb/presto#9004]
We ran a query in  Presto and seeing this error. Any idea how to deal with this issue? We are on the latest version of presto - 0.184.

`Query 20170919_173255_00010_hhkre failed: Encountered too many errors talking to a worker node. The node may have crashed or be under too much load. This is probably a transient issue, so please retry your query in a few minutes. (http://<MASKED_IP>1/v1/task/20170919_173255_00010_hhkre.4.4/results/5/0 - 6 failures, time since last success 60.85s)
`
And the logs say this - 

```
2017-09-19T17:34:16.455Z        INFO    Code-Cache-GC-Trigger   com.facebook.presto.server.CodeCacheGcTrigger   Triggering GC to avoid Code Cache eviction bugs
2017-09-19T17:34:29.008Z        ERROR   async-http-response-162 org.glassfish.jersey.server.ServerRuntime$Responder     An I/O error has occurred while writing a response message entity to the container output stream.
org.glassfish.jersey.server.internal.process.MappableException: org.eclipse.jetty.io.EofException
        at org.glassfish.jersey.server.internal.MappableExceptionWrapperInterceptor.aroundWriteTo(MappableExceptionWrapperInterceptor.java:92)
        at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:162)
        at org.glassfish.jersey.message.internal.MessageBodyFactory.writeTo(MessageBodyFactory.java:1130)
        at org.glassfish.jersey.server.ServerRuntime$Responder.writeResponse(ServerRuntime.java:711)
        at org.glassfish.jersey.server.ServerRuntime$Responder.processResponse(ServerRuntime.java:444)
        at org.glassfish.jersey.server.ServerRuntime$Responder.process(ServerRuntime.java:434)
        at org.glassfish.jersey.server.ServerRuntime$AsyncResponder$3.run(ServerRuntime.java:934)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
        at org.glassfish.jersey.server.ServerRuntime$AsyncResponder.resume(ServerRuntime.java:966)
        at org.glassfish.jersey.server.ServerRuntime$AsyncResponder.resume(ServerRuntime.java:922)
        at io.airlift.http.server.AsyncResponseHandler$1.onSuccess(AsyncResponseHandler.java:102)
        at com.google.common.util.concurrent.Futures$4.run(Futures.java:1135)
        at io.airlift.concurrent.BoundedExecutor.drainQueue(BoundedExecutor.java:78)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.eclipse.jetty.io.EofException
        at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:199)
        at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:420)
        at org.eclipse.jetty.io.WriteFlusher.completeWrite(WriteFlusher.java:375)
        at org.eclipse.jetty.io.SelectChannelEndPoint$3.run(SelectChannelEndPoint.java:107)
        at org.eclipse.jetty.io.SelectChannelEndPoint.onSelected(SelectChannelEndPoint.java:193)
        at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.processSelected(ManagedSelector.java:283)
        at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:181)
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:249)
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148)
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589)
        ... 1 more
Caused by: java.io.IOException: Broken pipe
        at sun.nio.ch.FileDispatcherImpl.write0(Native Method)
        at sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
        at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:93)
        at sun.nio.ch.IOUtil.write(IOUtil.java:65)
        at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:471)
        at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:177)
        ... 12 more

```

**Comment by nezihyigitbasi** [comment]
It's hard to tell what's happening by only looking at this log snippet. But, given this info I suspect like the full GC (garbage collection) right around the network operation may have caused it. In the logs I see

```
2017-09-19T17:34:16.455Z        INFO    Code-Cache-GC-Trigger   com.facebook.presto.server.CodeCacheGcTrigger   Triggering GC to avoid Code Cache eviction bugs
```

Which means full GC is triggered (unless you disable explicit full gcs in your jvm args) since the jvm code cache is close to full (this is some piece of code added to work around some jvm bug several releases ago). It's possible that the full GC pause may be causing this. Again to be sure more info and more debugging is needed.

Can you try bumping up `code-cache-collection-threshold` in your config to `70` to reduce the likelihood of such full gcs being triggered and see it helps?  (I have an open PR to bump up the current default value of 40 to 70 #9005)

**Comment by tomerstar** [comment]
Hi,
I encounter the same exact issue today using:
CDH 5.11.0,
Hive connetor
Presto 0.171
ORC partitioned table.

in case you run on similar env-
we saw that the crash was consistant for a specific Partition in our query.
We resolved it by rebuild the partition using Hive and afterwards the select was fine in presto.

I suspect, but cant verify, if this is somehow related to https://issues.apache.org/jira/browse/HIVE-11592.
the reason i suspect that is that we had this issue on this table when we built it lately, and only after rebuilding the specific orc table partitions that had this issue it was resolved.
maybe presto experts here will know better.
Tomer
